File: robin/python/level1/study2/press_keypad.py
def solution(numbers, hand):
    keypad = {1: [0, 0], 2: [0, 1], 3: [0, 2],
              4: [1, 0], 5: [1, 1], 6: [1, 2],
              7: [2, 0], 8: [2, 1], 9: [2, 2],
              '*': [3, 0], 0: [3, 1], '#': [3, 2]}

    answer = ''

    left = keypad.get('*')
    right = keypad.get('#')

    for i in numbers:

        key = keypad.get(i)

        left_distance = abs(left[0] - key[0]) + abs(left[1] - key[1])
        right_distance = abs(right[0] - key[0]) + abs(right[1] - key[1])

        if i in (1, 4, 7):
            left = key
            answer += "L"

        if i in (3, 6, 9):
            right = key
            answer += "R"

        if i in (2, 5, 8, 0):

            if left_distance == right_distance:
                if hand == "left":
                    left = key
                    answer += "L"
                else:
                    right = key
                    answer += "R"
            else:
                if left_distance < right_distance:
                    left = key
                    answer += "L"
                else:
                    right = key
                    answer += "R"

    return answer


if __name__ == '__main__':
    numbers = [1, 3, 4, 5, 8, 2, 1, 4, 5, 9, 5]
    hand = "right"
    print(solution(numbers, hand))
    # result = "LRLLLRLLRRL"

File: robin/java/src/level1/study2/PressKeypad.java
package level1.study2;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.List;

public class PressKeypad {

    public static String solution(int[] numbers, String hand) {
        String answer = "";

        /* 공식 ((|a - b|)/3  + (|a - b|)%3 ))
         * 1     ,2     ,3
         * 4     ,5     ,6
         * 7     ,8     ,9
         * 10(*) ,11(0) ,12(#)
         * */

        int leftIdx = 10;
        int rightIdx = 12;


        for (int number : numbers) {

            if (number == 0) number = 11;

            int leftDistance = Math.abs(leftIdx - number) / 3 + Math.abs(leftIdx - number) % 3;
            int rightDistance = Math.abs(rightIdx - number) / 3 + Math.abs(rightIdx - number) % 3;

            if (number == 1 || number == 4 || number == 7) {
                leftIdx = number;
                answer += "L";
            }

            if (number == 3 || number == 6 || number == 9) {
                rightIdx = number;
                answer += "R";
            }

            if (number == 2 || number == 5 || number == 8 || number == 11) {

                if (leftDistance == rightDistance) {

                    if (hand.equals("left")) {
                        leftIdx = number;
                        answer += "L";
                    }

                    if (hand.equals("right")) {
                        rightIdx = number;
                        answer += "R";
                    }

                } else {

                    if (leftDistance < rightDistance) {
                        leftIdx = number;
                        answer += "L";
                    }

                    if (leftDistance > rightDistance) {
                        rightIdx = number;
                        answer += "R";
                    }
                }
            }
        }
        return answer;
    }

    public static void main(String[] args) {
        int[] numbers = {1, 3, 4, 5, 8, 2, 1, 4, 5, 9, 5};
        String hand = "right";
        System.out.println(solution(numbers, hand));
        //LRLLLRLLRRL
    }
}
